### PR TITLE
[Enhancement] Make rewriter more powerful

### PR DESCRIPTION
**Pull request by SingleZombie**
## Motivation

1. We need to select IR in rewriters.
2. We want to check the version of libraries in rewriters

## Modification

1. Add utils about the environment and modify check_env.py
2. Modify the implementation of the rewriter registry.
3. Modify the related interfaces.
4. Add unit tests.
5. Add docstring.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

@FUNCTION_REWRITER.register_rewriter(
    func_name='mmdeploy.codebase.mmdet.deploy.utils.clip_bboxes',
    backend='tensorrt',
    extra_checkers=LibVersionChecker('tensorrt', min_version='8'))
def clip_bboxes__trt8(x1: Tensor, y1: Tensor, x2: Tensor, y2: Tensor,
                      max_shape: Union[Tensor, Sequence[int]]):
...

### CODE REVIEW

File: tests/test_core/test_rewriter_utils.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+def test_collect_env():
+    env_dict = collect_env(Backend.ONNXRUNTIME, IR.ONNX, version='1.0')
+    assert env_dict['backend'] == Backend.ONNXRUNTIME
+    assert env_dict['ir'] == IR.ONNX
+    assert env_dict['version'] == '1.0'
+    assert env_dict['mmdeploy'] == mmdeploy.__version__
+
+
+class TestCheker:
```
Comment (RunningLeon) [inline tests/test_core/test_rewriter_utils.py:18]: typo

File: mmdeploy/utils/env.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    Args:
+        lib (str): The name of library.
+
+    Returns:
+        None | str: If the library has been installed, return version.
+    """
+    try:
+        lib = importlib.import_module(lib)
+    except ImportError:
```
Comment (grimoire) [inline mmdeploy/utils/env.py:18]: load library might throw other exception.
Such as mmocr will assert the mmcv verison. which will thown an assert error.

File: mmdeploy/core/rewriters/rewriter_utils.py [hunk 3]
```diff
@@ -78,55 +141,121 @@ class RewriterRegistry:
...
+            default_rewriter = None
+            final_rewriter = None
+            for record in rewriter_records:
+                # Get the checkers of current rewriter
+                checkers: List[Checker] = record['_checkers']
+
+                # Check if the rewriter is default rewriter
+                if len(checkers) == 0:
+                    #  Process the default rewriter exceptionally
+                    default_rewriter = record
```
Comment (grimoire) [inline mmdeploy/core/rewriters/rewriter_utils.py:176]: Should we add a warning if multiple default_rewriters were found?

**Review by AllentDan — COMMENTED** [review]
Deploying Faster-RCNN to TensorRT gives me the following errors:
```
TypeError: forward() missing 1 required positional argument: 'img_metas'
2022-02-14 16:12:39,950 - mmdeploy - ERROR - torch2onnx failed.
```

**Comment by RunningLeon** [comment]
@SingleZombie docstring coverage drops below threshold.

File: mmdeploy/core/rewriters/function_rewriter.py [hunk 4]
```diff
@@ -66,32 +67,30 @@ class FunctionRewriter:
...
+            backend: str = Backend.DEFAULT.value,
+            ir: IR = IR.DEFAULT,
+            extra_checkers: Optional[Union[Checker, List[Checker]]] = None,
+            **kwargs):
         """The interface of function rewriter decorator.
 
         Args:
             func_name (str): The function name/path to rewrite.
-            backend (str): The inference engine name.
+            backend (str): The rewriter will be activated on which backend.
```
Comment (grimoire) [inline mmdeploy/core/rewriters/function_rewriter.py:81]: Add docstring aboult ir and extra_checker

**Comment by SingleZombie** [comment]
> @SingleZombie docstring coverage drops below threshold.

Yes. This PR is labeled with WIP. I will remove the WIP as soon as I update all docstring.

File: mmdeploy/core/rewriters/rewriter_utils.py [hunk 5]
```diff
@@ -56,6 +59,66 @@ def import_function(path: str) -> Tuple[Callable, Optional[type]]:
...
+    def __init__(self, lib: str, min_version=None, max_version=None):
+        super().__init__()
+        self.lib = lib
+        self.min_version = min_version
+        self.max_version = max_version
+
+    def check(self, env: Dict) -> bool:
+        from packaging import version
+        valid = True
+        if self.min_version is not None:
```
Comment (RunningLeon) [inline mmdeploy/core/rewriters/rewriter_utils.py:113]: logic is not good enough. For instance, if min_version is not met and max_version is met, then final `valid=True` while actually the min_version is not met.

Comment (SingleZombie) [reply mmdeploy/core/rewriters/rewriter_utils.py:113]: Yes! The codes are wrong. I will fix them!

File: mmdeploy/core/rewriters/rewriter_manager.py [hunk 6]
```diff
@@ -84,20 +76,20 @@ class RewriterContext:
...
                  cfg: Dict = dict(),
                  backend: str = Backend.DEFAULT.value,
+                 ir: IR = IR.DEFAULT,
                  rewriter_manager: RewriterManager = REWRITER_MANAGER,
                  **kwargs):
         self._cfg = cfg
-        self._backend = backend
         self._kwargs = kwargs
         self._rewriter_manager = rewriter_manager
+        # TODO: Make the type of parameter backend to Backend
```
Comment (RunningLeon) [inline mmdeploy/core/rewriters/rewriter_manager.py:85]: Is this todo finished in this PR?

File: mmdeploy/core/rewriters/module_rewriter.py [hunk 7]
```diff
@@ -67,7 +74,9 @@ def patch_model(self,
             >>> patched_model = patch_model(model, cfg=deploy_cfg,
             >>>                             backend=backend)
         """
-        self._collect_record(backend)
+        # TODO: Make the type of parameter backend to Backend
```
Comment (RunningLeon) [inline mmdeploy/core/rewriters/module_rewriter.py:77]: could we remove this comment?

File: mmdeploy/core/rewriters/rewriter_manager.py (hunk 6 above)
Comment (SingleZombie) [reply mmdeploy/core/rewriters/rewriter_manager.py:85]: No. I think we can do it in the future.

File: mmdeploy/core/rewriters/module_rewriter.py (hunk 7 above)
Comment (SingleZombie) [reply mmdeploy/core/rewriters/module_rewriter.py:77]: Yes

**Review by grimoire — APPROVED** [review]

**Review by RunningLeon — DISMISSED** [review]
LGTM

**Comment by RunningLeon** [comment]
Suggest to merge this PR after release of v0.3.0.

**Review by RunningLeon — APPROVED** [review]

**Review by grimoire — APPROVED** [review]

**Pull request merged**